### PR TITLE
28 extraction statistical parameters

### PR DESCRIPTION
**Pull request by Tole-k**
my idea of statistical parameters extraction, if there are some other parameters you would like to see, please let me know

### CODE REVIEW

**Review by Dawid64 — CHANGES_REQUESTED** [review]
There are few things to change and few to consider:
Cosmetic:
- `from sklearn.preprocessing import OneHotEncoder as sklearnOneHotEncoder`, so our class can be called properly

Usage:
- I think it would be the best to set device in utils options, it's less likely for casual usage to change in the middle of program (I will make commit with this change, you can change it if you find better approach)
- It may be easier for future to create dataclass to contain information, it may be easier to remember all fields with easy-to-find way, rather than using dictionary (of course it behave similarly to dictionary `dataclass_object['mean'] = 10`
- As we want to use neural network based on this parameters it's important to have finite value of them, not dependent on other features as number of columns
- Found an error, the values of quantiles should be different, not sure why it occurred, you can check
Screenshot 2024-09-02 at 11 46 43 AM
image

- Also I believe we shouldn't normalize the original dataset, statistical parameter extractor should just extract parameters, not manipulate the dataset

I think we should focus mostly on parametrizing whole dataset rather than specific columns, so it's easier to create fixed-size neural network, here are some ideas:
- num_columns
- num_rows
- num of target classes (for classification problems)
- size of the biggest class (for classification problems)
- size of the smallest class (for classification problems)
- feature dependencies (value based on dependency matrix)

**Comment by Tole-k** [comment]
> * feature dependencies (value based on dependency matrix)
 
You mean correlation matrix, right? Wouldn't that be dependent on the number of columns, so the opposite what you mentioned earlier?

**Comment by Dawid64** [comment]
> You mean correlation matrix, right? Wouldn't that be dependent on the number of columns, so the opposite what you mentioned earlier?

Yes, I meant correlation matrix, forgot the name, sorry for that, I meant some values describing matrix, as 'num_high/low_values' or maybe even highest / lowest eigenvalues may tell something, as correlation matrix contains important information about connection between different features of the dataset, taking part of them may have positive result

**Comment by Dawid64** [comment]
> > * Also, I believe we shouldn't normalize the original dataset, statistical parameter extractor should just extract parameters, not manipulate the dataset
> 
> Do you mean my usage of preprocessing methods inside the main bloc? that was only for testing purposes, to verify whether extraction works with preprocessed datasets. The Extractor class itself does return only the description and does not transform the dataset in any way

You're right, my bad, then you may think about adding some normalization into parameters, as for example std is dependent on that, but ofc without manipulating dataset

**Comment by Dawid64** [comment]
I made few changes, mostly added type hints, changed classes into data classes (as far as I know their may be more efficient methods, but it should do and it's also good practice) and moved import, as in library we may not provide datasets directory, at least not in this form, if they are incorrect let me know I will remove the commit (sorry for not doing pull request into this branch, forgot about this option).
From other changes we may think about connecting target and data description into one, as we want to pass it to neural network (or other algorithm), so it will be the best to store it in one place.
Also the name of the biggest class probably will not be important, and rather than sizes in values, we may express it in % of the whole dataset (I think it will be more helpful, as it will be normalized, and size is already known) we laso may think about num of lowly/highly correlated features (if to normalized them or not)

**Comment by Tole-k** [comment]
> I made few changes, mostly added type hints, changed classes into data classes (as far as I know their may be more efficient methods, but it should do and it's also good practice) and moved import, as in library we may not provide datasets directory, at least not in this form, if they are incorrect let me know I will remove the commit (sorry for not doing pull request into this branch, forgot about this option). From other changes we may think about connecting target and data description into one, as we want to pass it to neural network (or other algorithm), so it will be the best to store it in one place. Also the name of the biggest class probably will not be important, and rather than sizes in values, we may express it in % of the whole dataset (I think it will be more helpful, as it will be normalized, and size is already known) we laso may think about num of lowly/highly correlated features (if to normalized them or not)

I've already added the number of lowly corelated features and the number of highly corelated features, unless you mean something different

**Comment by Dawid64** [comment]
> I've already added the number of lowly corelated features and the number of highly corelated features, unless you mean something different
I meant either to normalize them, as % of total, or not sure which will be more universal
